### ISSUE TITLE
project verification 

### ISSUE THREAD

**Issue by zacharylaguna** [MyMathApp-Fall2023/OnlineTextbook#2]
Our Team: Capstone Matching
Other Team: My Math Apps

THEIR REPO WAS PRIVATE, SO WE DID NOT HAVE ACCESS UNTIL THE END OF THE CLASS

Documentation Covers Getting Started:
Specify which nodejs version (just mentioning “latest” is not okay as the project is a legacy project)
Npm is not automatically installed with nodejs (just tell us the commands to install them)
Add sudo in front of install command
No documentation for installing Nest
Cannot install matching version of Yarn to nodejs
For the create database mymathapps command, we needed to use back quotes instead of quotes
Vague instructions. Specify the commands and how to do them.
Install nodejs (it will automatically install npm)
Install Docker and set it up. 
Copy the init.sql file into the Docker container, run it, then run the tests.
.env.example file in the server folder was a wrong (configuration?) file example. We had to spend considerable time debugging it.
The line should be only
DB_Host = "localhost"

and not
DB_Host = "localhost"  // followed by the comments

How to run tests:
Documentation is available for testing
Instructions are not clear. Specifically, “Copy the init.sql file into the Docker container, run it, then run the tests.”

Testing:
Server-side
Test Suites: 2 failed, 9 passed, 11 total
Tests:   	8 skipped, 60 passed, 68 total

Client-side
Got this error

yarn run v1.22.21
warning ../package.json: No license field
warning ../../package.json: No license field
$ node_modules/.bin/cucumber-js -f @cucumber/pretty-formatter --publish-quiet src/test/features/
This Node.js version (v21.2.0) has not been tested with this version of Cucumber; it should work normally, but please raise an issue if you see anything unexpected.
VError: a BeforeAll hook errored, process exiting: src/test/features/step-definitions/SubscriptionPage.js:9: Cannot read properties of null (reading 'get')
	at Runtime.runTestRunHooks (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/runtime/run_test_run_hooks.js:23:23)
	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
	at async Runtime.start (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/runtime/index.js:53:9)
	at async runCucumber (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/api/run_cucumber.js:106:21)
	at async Cli.run (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/cli/index.js:50:29)
	at async Object.run [as default] (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/cli/run.js:34:18)
caused by: TypeError: Cannot read properties of null (reading 'get')
	at /home/yash/OnlineTextbook/client/src/test/features/step-definitions/SubscriptionPage.js:12:18
	at Object.run (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/user_code_runner.js:24:27)
	at Runtime.runTestRunHooks (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/runtime/run_test_run_hooks.js:15:64)
	at Runtime.start (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/runtime/index.js:53:20)
	at runCucumber (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/api/run_cucumber.js:106:35)
	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
	at async Cli.run (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/cli/index.js:50:29)
	at async Object.run [as default] (/home/yash/OnlineTextbook/client/node_modules/@cucumber/cucumber/lib/cli/run.js:34:18)
error Command failed with exit code 1.

Deployment:
Documentation is available for deployment
Error at step 3
heroku git:remote -a example-app
 ›   Error: You do not have access to the app example-app.
 ›
 ›   Error ID: forbidden

The documentation assumes that the user has already deployed to Heroku before or has some knowledge of it. If we blindly follow the documentation as a new user, it is hard to follow and understand.
Vague instructions. Mention the commands as well
Create a local branch off of dev
Modify the .gitignore to allow .env.heroku to be permitted
Etc

Deployed app at https://mymathapp-verification-b89c543735f3.herokuapp.com/. Got this deployment errors

2023-11-20T22:02:37.967524+00:00 heroku[web.1]: Starting process with command `cd server && npm run start:heroku`
2023-11-20T22:02:38.739468+00:00 app[web.1]:
2023-11-20T22:02:38.739502+00:00 app[web.1]: > mymath-server@2.5.3 start:heroku
2023-11-20T22:02:38.739503+00:00 app[web.1]: > cross-env RUNNING_ENV=heroku node dist/main
2023-11-20T22:02:38.739503+00:00 app[web.1]:
2023-11-20T22:02:39.539632+00:00 app[web.1]: /app/server/dist/app.module.js:78
2023-11-20T22:02:39.539652+00:00 app[web.1]: synchronize: process.env.ENV_TYPE.toUpperCase() === 'HEROKU',
2023-11-20T22:02:39.539653+00:00 app[web.1]: ^
2023-11-20T22:02:39.539657+00:00 app[web.1]:
2023-11-20T22:02:39.539658+00:00 app[web.1]: TypeError: Cannot read properties of undefined (reading 'toUpperCase')
2023-11-20T22:02:39.539658+00:00 app[web.1]: at Object.<anonymous> (/app/server/dist/app.module.js:78:51)
2023-11-20T22:02:39.539658+00:00 app[web.1]: at Module._compile (node:internal/modules/cjs/loader:1119:14)
2023-11-20T22:02:39.539658+00:00 app[web.1]: at Module._extensions..js (node:internal/modules/cjs/loader:1173:10)
2023-11-20T22:02:39.539658+00:00 app[web.1]: at Module.load (node:internal/modules/cjs/loader:997:32)
2023-11-20T22:02:39.539659+00:00 app[web.1]: at Module._load (node:internal/modules/cjs/loader:838:12)
2023-11-20T22:02:39.539659+00:00 app[web.1]: at Module.require (node:internal/modules/cjs/loader:1021:19)
2023-11-20T22:02:39.539659+00:00 app[web.1]: at require (node:internal/modules/cjs/helpers:103:18)
2023-11-20T22:02:39.539659+00:00 app[web.1]: at Object.<anonymous> (/app/server/dist/main.js:4:22)
2023-11-20T22:02:39.539659+00:00 app[web.1]: at Module._compile (node:internal/modules/cjs/loader:1119:14)
2023-11-20T22:02:39.539660+00:00 app[web.1]: at Module._extensions..js (node:internal/modules/cjs/loader:1173:10)
2023-11-20T22:02:39.539660+00:00 app[web.1]:
2023-11-20T22:02:39.539660+00:00 app[web.1]: Node.js v18.9.0
2023-11-20T22:02:39.668380+00:00 heroku[web.1]: Process exited with status 1
2023-11-20T22:02:39.686818+00:00 heroku[web.1]: State changed from starting to crashed
2023-11-20T22:02:39.690486+00:00 heroku[web.1]: State changed from crashed to starting
2023-11-20T22:02:57.510320+00:00 heroku[web.1]: Starting process with command `cd server && npm run start:heroku`
2023-11-20T22:02:58.398741+00:00 app[web.1]:
2023-11-20T22:02:58.398761+00:00 app[web.1]: > mymath-server@2.5.3 start:heroku
2023-11-20T22:02:58.398761+00:00 app[web.1]: > cross-env RUNNING_ENV=heroku node dist/main
2023-11-20T22:02:58.398761+00:00 app[web.1]:
2023-11-20T22:02:59.263926+00:00 app[web.1]: /app/server/dist/app.module.js:78
2023-11-20T22:02:59.263952+00:00 app[web.1]: synchronize: process.env.ENV_TYPE.toUpperCase() === 'HEROKU',
2023-11-20T22:02:59.263954+00:00 app[web.1]: ^
2023-11-20T22:02:59.263954+00:00 app[web.1]:
2023-11-20T22:02:59.263955+00:00 app[web.1]: TypeError: Cannot read properties of undefined (reading 'toUpperCase')
2023-11-20T22:02:59.263955+00:00 app[web.1]: at Object.<anonymous> (/app/server/dist/app.module.js:78:51)
2023-11-20T22:02:59.263955+00:00 app[web.1]: at Module._compile (node:internal/modules/cjs/loader:1119:14)
2023-11-20T22:02:59.263955+00:00 app[web.1]: at Module._extensions..js (node:internal/modules/cjs/loader:1173:10)
2023-11-20T22:02:59.263956+00:00 app[web.1]: at Module.load (node:internal/modules/cjs/loader:997:32)
2023-11-20T22:02:59.263956+00:00 app[web.1]: at Module._load (node:internal/modules/cjs/loader:838:12)
2023-11-20T22:02:59.263956+00:00 app[web.1]: at Module.require (node:internal/modules/cjs/loader:1021:19)
2023-11-20T22:02:59.263956+00:00 app[web.1]: at require (node:internal/modules/cjs/helpers:103:18)
2023-11-20T22:02:59.263957+00:00 app[web.1]: at Object.<anonymous> (/app/server/dist/main.js:4:22)
2023-11-20T22:02:59.263957+00:00 app[web.1]: at Module._compile (node:internal/modules/cjs/loader:1119:14)
2023-11-20T22:02:59.263957+00:00 app[web.1]: at Module._extensions..js (node:internal/modules/cjs/loader:1173:10)
2023-11-20T22:02:59.263958+00:00 app[web.1]:
2023-11-20T22:02:59.263958+00:00 app[web.1]: Node.js v18.9.0
2023-11-20T22:02:59.356199+00:00 heroku[web.1]: Process exited with status 1
2023-11-20T22:02:59.377059+00:00 heroku[web.1]: State changed from starting to crashed
2023-11-20T22:03:33.334069+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=mymathapp-verification-b89c543735f3.herokuapp.com request_id=bbc00065-6e8a-464b-ae65-5217629d8c84 fwd="75.108.161.130" dyno= connect= service= status=503 bytes= protocol=https
2023-11-20T22:04:01.000000+00:00 app[api]: Build succeeded

Not able to “Modify the .gitignore to allow .env.heroku to be permitted”
Not proceeding any further.

Problems:
Could not get running because of version dependencies (fixed)
‘Yarn start’ resulted in the error: getaddr info ENOTFOUND localhost - not able to connect to the database (could not fix) prevented us from being able to run the application (fixed)

Contact Info:
Contact info is available for questions and troubleshooting

**Comment by smbanasik** [comment]
Documentation has been improved and reviewed by 2 other members of team.
